### PR TITLE
Abstracted auth, now allows auth disabled

### PR DESCRIPTION
**Pull request by jilleJr**
- \[x] I've added a new note in the `CHANGELOG.md` file, according to docs:
  https://iver-wharf.github.io/#/development/changelogs/writing-changelogs

## Summary

- Duplicated auth module and components into:

  - src/app/auth/auth-fake: Just returns `isAuthenticated=true`
  - src/app/auth/auth-oidc: Performs OIDC logic

- Changed above new modules into lazy-loaded modules.

- Added `src/assets/fake-auth/.well-known/openid-configuration.json` so trick the OIDC library that it's valid when we don't want it.

- Changed to auto-login in OIDC auth implementation.

## Motivation

Want to be able to NOT use authentication locally, and in some installations of Wharf.

A lot of trial and error. I did not have a clue of how to make this work, so a lot of stuff was tried out.

I tried but failed to lazy-load the external OIDC library. It **has** to be loaded in at the root app module, because *of reasons*... But that's OK because I found a way to cheat around it when we don't want it: by adding the `src/assets/fake-auth/.well-known/openid-configuration.json`.

### CODE REVIEW

**Comment by jilleJr** [comment]
The failing CI step is by Trivy (Docker image scanning). I've opened a separate PR for that: #139

File: src/app/auth/auth-oidc-config/oidc-auth-config.module.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { HTTP_INTERCEPTORS } from '@angular/common/http';
+import { NgModule } from '@angular/core';
+import {
+  AuthModule,
+} from 'angular-auth-oidc-client';
```
Comment (Alexamakans) [inline src/app/auth/auth-oidc-config/oidc-auth-config.module.ts:5]:
```suggestion
import { AuthModule } from 'angular-auth-oidc-client';
```

File: src/app/auth/auth-oidc-config/oidc-auth.interceptor.ts [hunk 2]
```diff
@@ -5,13 +5,16 @@ import { Observable } from 'rxjs';
...
     private oidcSecurityService: OidcSecurityService,
   ) { }
 
   intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    if (!environment.oidcConfig?.enabled) {
+      return next.handle(req);
+    }
     const apiUrl = environment.backendUrls.api;
     if (!req.url.includes(apiUrl)) {
       return next.handle(req);
```
Comment (Alexamakans) [inline src/app/auth/auth-oidc-config/oidc-auth.interceptor.ts:20]: We probably never want to intercept URLs that don't _start with_ the `apiUrl`.
I believe we had a short discussion about this before, so pointing it out again:

Probably best to change to using `if (!req.url.startsWith(apiUrl)) {` here, just in case.

File: src/app/auth/auth-oidc/oidc-auto-login.guard.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+          this.authService.navigateBackToReturnUrl();
+        } else {
+          this.authService.setReturnUrl(url);
+          this.oidcAuthService.authorize();
+        }
+
+        return isAuthenticated;
+      }),
+    );
+
```
Comment (Alexamakans) [inline src/app/auth/auth-oidc/oidc-auto-login.guard.ts:69]:
```suggestion
```

File: src/app/auth/auth-oidc/oidc-home/oidc-home.component.ts [hunk 4]
```diff
@@ -27,9 +26,14 @@ export class LoginComponent implements OnInit {
...
-    this.configuration = this.oidcSecurityService.getConfiguration();
+    this.configuration = environment.oidcConfig;
     this.userData$ = this.oidcSecurityService.userData$;
-    this.isAuthenticated$ = this.oidcSecurityService.isAuthenticated$.pipe(pluck('isAuthenticated'));
+    this.isAuthenticated$ = this.oidcSecurityService.isAuthenticated$
+      .pipe(pluck('isAuthenticated'));
+
+    this.oidcSecurityService.isAuthenticated$.subscribe({
+      next: b => console.log('is authenticated? ', b),
+    });
```
Comment (Alexamakans) [inline src/app/auth/auth-oidc/oidc-home/oidc-home.component.ts:36]: Shouldn't this be using the plucked version?
```suggestion
    this.isAuthenticated$.subscribe({
      next: b => console.log('is authenticated? ', b),
    });
```

File: src/app/auth/auth.guard.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    return this.checkAuth(state.url);
+  }
+
+  private checkAuth(url: string) {
+    if (this.authService.isAuthenticated) {
+      // logged in so return true
+      return true;
+    }
+
+    // not logged in so redirect to login page with the return url and return false
```
Comment (Alexamakans) [inline src/app/auth/auth.guard.ts:28]: nit: Personally don't think the comments in this method are necessary, but if keeping them then I suggest rewording slightly.
```suggestion
    // not logged in so set the return url, redirect to the login page, and return false
```

**Review by Alexamakans — CHANGES_REQUESTED** [review]
Looked over mostly for weird things and errors/problems that don't span several files.
If it has been tested and it works then I'd be willing to approve. Good work 👍🏻

File: Makefile [hunk 6]
```diff
@@ -32,7 +32,7 @@ ifneq "$(version)" "latest"
 endif
 
 docker-run:
-	docker run --rm -it quay.io/iver-wharf/wharf-web:$(version)
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply Makefile:35]: Was missing the port exposing. The `make docker-run` started the server alright, but you couldn't access it at http://localhost:8080

File: src/app/auth/auth-oidc-config/oidc-auth.interceptor.ts (hunk 2 above)
Comment (jilleJr) [reply src/app/auth/auth-oidc-config/oidc-auth.interceptor.ts:20]: Yea, totally agree.

I've used your suggestion, and also added checks if it's one of the importer's URLs as well.

File: src/app/auth/auth-oidc/oidc-home/oidc-home.component.ts (hunk 4 above)
Comment (jilleJr) [reply src/app/auth/auth-oidc/oidc-home/oidc-home.component.ts:36]: No, it's testing stuff. It shouldn't be there at all :)

File: src/app/auth/auth.guard.ts (hunk 5 above)
Comment (jilleJr) [reply src/app/auth/auth.guard.ts]: I've removed the comments.

**Review by Alexamakans — APPROVED** [review]
LGTMGTM 👍🏻
